src/hotspot/share/opto/parse3.cpp

```diff
@@ -256,10 +255,9 @@ void Parse::do_put_xxx(Node* obj, ciField* field, bool is_field) {
       return;
     }
   }
-  if (obj->is_InlineType()) {
-    set_inline_type_field(obj, field, val);
-    return;
-  }
+  val = cast_non_larval(val);
+
+  assert(!obj->is_InlineType(), "InlineTypeNodes are non-larval value objects");
```

Comment:
The assert message is a bit confusing. Maybe improve it or add a comment that the field holder should not have been scalarized because it's a larval.